src/hotspot/share/gc/parallel/psParallelCompact.hpp

```diff
@@ -116,51 +116,45 @@ class SplitInfo
   // recorded).  The very first region cannot have a partial object and thus is
   // never split, so 0 is the 'invalid' value.
-  bool is_valid() const { return _src_region_idx > 0; }
+  bool is_valid() const { return _split_region_idx > 0; }
 
   // Return true if this split holds data for the specified source region.
-  inline bool is_split(size_t source_region) const;
-
-  // The index of the split region, the size of the partial object on that
-  // region and the destination of the partial object.
-  size_t    partial_obj_size() const { return _partial_obj_size; }
-  HeapWord* destination() const      { return _destination; }
+  inline bool is_split(size_t region_idx) const;
+
+  // A split region has two "destinations", living in two spaces. This method
+  // returns the split (first) one -- destination for the first live word on
+  // this split region.
+  HeapWord* split_destination() const {
+    assert(_split_destination != nullptr, "inv");
+    return _split_destination;
+  }
 
-  // The destination count of the partial object referenced by this split
-  // (either 1 or 2).  This must be added to the destination count of the
-  // remainder of the source region.
-  unsigned int destination_count() const { return _destination_count; }
+  // Obj at the split point doesn't fit the previous space and will be relocated to the next space.
+  HeapWord* split_point() const { return _split_point; }
 
-  // If a word within the partial object will be written to the first word of a
-  // destination region, this is the address of the destination region;
-  // otherwise this is null.
-  HeapWord* dest_region_addr() const     { return _dest_region_addr; }
+  // Number of live words before the split point on this region.
+  size_t preceding_live_words() const { return _preceding_live_words; }
 
-  // If a word within the partial object will be written to the first word of a
-  // destination region, this is the address of that word within the partial
-  // object; otherwise this is null.
-  HeapWord* first_src_addr() const       { return _first_src_addr; }
+  // Number of regions the preceding live words are relocated into.
+  uint split_destination_count() const { return _split_destination_count; }
 
-  // Record the data necessary to split the region src_region_idx.
-  void record(size_t src_region_idx, size_t partial_obj_size,
-              HeapWord* destination);
+  void record(size_t split_region_idx, HeapWord* split_point, size_t preceding_live_words);
 
   void clear();
 
   DEBUG_ONLY(void verify_clear();)
 
 private:
-  size_t       _src_region_idx;
-  size_t       _partial_obj_size;
-  HeapWord*    _destination;
-  unsigned int _destination_count;
-  HeapWord*    _dest_region_addr;
-  HeapWord*    _first_src_addr;
+  size_t       _split_region_idx;
+  HeapWord*    _split_destination;
+  HeapWord*    _split_point;
+  size_t       _preceding_live_words;
+  uint         _split_destination_count;
```

Comment:
I tried to make fields of `class SplitInfo` to match their counterpart in `class RegionData`, except the additional `_split_` prefix, in order to signify these fields are closely related.

> But in _split_destination_count, the prefix split means two parts whose destinations locate in different regions but in the same space.

Why are they in the "same" space? The purpose of having "split" to support space-boundary so that the first part and the second part are in two spaces.

Reply:
> > But in _split_destination_count, the prefix split means two parts whose destinations locate in different regions but in the same space.
> 
> Why are they in the "same" space? The purpose of having "split" to support space-boundary so that the first part and the second part are in two spaces.

The `_split_destination_count` is a size to record how many regions the preceding part occupy. The preceding part is moved to one same space and the second part is moved to another space. So now, it seems you are also misled by this ambiguous name, which is a proof that we need to rename it.

```
// file psParallelCompact.hpp

  // Number of regions the preceding live words are relocated into.
  uint split_destination_count() const { return _split_destination_count; }


// file psParallelCompact.cpp, method SplitInfo::record

  // How many regions does the preceding part occupy
  uint split_destination_count;
  if (preceding_live_words == 0) {
    split_destination_count = 0;
  } else {
    if (split_destination + preceding_live_words > sd.region_align_up(split_destination)) {
      split_destination_count = 2;
    } else {
      split_destination_count = 1;
    }
  }
```

Reply:
Since the preceding live words can be relocated to more than one region, I believe you interpret that as "split". However, the word "split" in this class/context is exclusively reserved for "splitting" a region so that "The preceding part is moved to one same space and the second part is moved to another space."

Relocating into multiple destination-regions is not "split", otherwise, `RegionData::_destination` should contain "split" as well.

Reply:
> In _split_destination, the prefix split means two parts whose destinations locate in different spaces. But in _split_destination_count, the prefix split means two parts whose destinations locate in different regions but in the same space.

The two `two parts` I mentioned is unclear and ambiguous. To be clearer and be easy to understand, please omit my previous unclear statement, and help judge whether my following understanding is right or not. Thanks.

One source region which needs to be split can be divided into three parts (shown below).
- `first-part-opt-1` contains objects targeted to one region
- `first-part-opt-2` contains objects targeted to another region (but the same space as the target of `first-part-opt-1`)
- `second-part` contains objects targeted to another space
- `first-part-opt-1` and `first-part-opt-2` are optional

```
source-region:
-----------------------------------------------------------
|  first-part-opt-1  |  first-part-opt-2  |  second-part  |
a--------------------b--------------------c---------------d
```

Then we have several conclusions:
- the `_split_point` is at the location `c`
- the `_preceding_live_words` is the size of the live objects from `a` to `c`
- the `_split_destination` is the destination of the first object from `a`(not the first object from `c`)
- the `_split_destination_count` has three conditions
  - if the `_preceding_live_words` is `0`, the `_split_destination_count` is `0`, because the `first-part-opt-1` and `first-part-opt-2` don't exist.
  -  if `_split_destination + _preceding_live_words` is larger than `region_align_up(_split_destination)`,  the `_split_destination_count` is `2`, because both `first-part-opt-1` and `first-part-opt-2` exist.
  - otherwise, the `_split_destination_count` is `1` (only `first-part-opt-1` exists)

> However, the word "split" in this class/context is exclusively reserved for "splitting" a region so that "The preceding part is moved to one same space and the second part is moved to another space."

If my understanding is right, I think the word `split` can be used in the context related to the `c` or `from c to else` or `second-part`. And the word `preceding` can be used in the context related to the `a`, `b` or `from 'a'/'b' to else` or `first-part-opt-1/first-part-opt-2`. What do you think about it? If you agree with this convention. I propose to change `_split_destination` to `_preceding_destination` and change `_split_destination_count` to `_preceding_destination_count`.